src/freenet/node/DarknetPeerNode.java

```diff
@@ -1393,12 +1395,23 @@ public int sendDownloadFeed(FreenetURI URI, String description) {
 
 	public int sendTextFeed(String message) {
 		long now = System.currentTimeMillis();
-		SimpleFieldSet fs = new SimpleFieldSet(true);
-		fs.put("type", Node.N2N_TEXT_MESSAGE_TYPE_USERALERT);
-		fs.putSingle("text", Base64.encodeUTF8(message));
-		fs.put("composedTime", now);
-		sendNodeToNodeMessage(fs, Node.N2N_MESSAGE_TYPE_FPROXY, true, now, true);
-		this.setPeerNodeStatus(System.currentTimeMillis());
+		long msgid = node.random.nextLong();
+		// split large messages
+		int requiredN2nCount = 1 + (message.length() / 1024);
+		String messagePart;
+		for (int i = 0; i < requiredN2nCount; i++) {
+			messagePart = message.substring(i * 1024,
+							Math.min((i+1) * 1024,
+								 message.length()));
+			SimpleFieldSet fs = new SimpleFieldSet(true);
+			fs.put("type", Node.N2N_TEXT_MESSAGE_TYPE_USERALERT);
```

Comment:
This will sound theoretical at first but it has a real consequence already which I will explain later on:
You changed the network protocol here without introduction a protocol version field to indicate a different version or at least changing the ```type``` value here to something different.
This means it's impossible for peers to recognize whether they're talking to something which speaks the old or the new protocol version and and act accordingly - which can cause trouble in handling cases (or even bugs) where the protocol version needs to be identified.

The already existing consequence is that the UserAlert hack which I already commented is necessary: You cannot detect whether a message is split until more messages with the same ID arrive. So the first piece is displayed before others arrive even though it should be hidden until the message is complete.

I accept your desire to have this stay backwards compatible by not changing the ```type``` so I would suggest the following incremental modification to be backwards compatible *and* versioned:
Add an integer ```fragments``` which contains the total number of 1024 byte pieces the message will consist of, and an integer ```fragment``` which identifies the current fragment. 
If those aren't present (or both equal to 1) in a received message then assume it's the old format of non-split messages.

Then putting the message together also becomes as easy as allocating a String array of the size being equal to ```fragments``` and putting the fragments into the index ```fragment```. Don't forget to add code to throw away all messages of the given ```msgid``` if malicious intent is indicated by e.g. an excessive ```fragments``` count, String size, multiple messages claiming a different ```fragments``` for a given ```msgid```, or multiple messages using the same ```fragment``` number. Fred probably even has code to disconnect/punish peers if they do such things.

Reply:
Ah sorry I didn't notice that ```msgid``` serves as an indicator for the new protocol :|
Still adding fragment numbers will be necessary to resolve the issue of being able to detect when a message is complete.

Reply:
Arguably the fact that I missed that ```msgid``` indicates the new protocol shows that it's a poor indicator of a message being split, it's not very intuitive. Normally either every object of a certain class has an identifier or the class doesn't have an identifier, selectively only adding an identifier to some is weird. An identifier doesn't sound very related to splitting, does it? :|

Having an ID in all, even non-split, messages will perhaps be useful for other things anyway such as sending read reciepts so using it for this meaning would block those things in the future.